homeassistant/components/lutron_caseta/__init__.py

```diff
@@ -91,17 +135,168 @@ async def async_setup_entry(hass, config_entry):
+        return
+
+    _LOGGER.debug("Connected to Lutron Caseta bridge via LIP at %s:23", host)
+    button_devices_by_lip_id = _async_merge_lip_leap_data(lip_devices, bridge)
+    button_devices_by_dr_id = await _async_register_button_devices(
+        hass, config_entry_id, bridge_device, button_devices_by_lip_id
+    )
+    _async_subscribe_pico_remote_events(hass, lip, button_devices_by_lip_id)
+    data[BUTTON_DEVICES] = button_devices_by_dr_id
+    data[BRIDGE_LIP] = lip
+
+
+@callback
+def _async_merge_lip_leap_data(lip_devices, bridge):
+    """Merge the leap data into the lip data."""
+    sensor_devices = bridge.get_devices_by_domain("sensor")
+
+    button_devices_by_id = {
+        id: device for id, device in lip_devices.items() if "Buttons" in device
+    }
+    sensor_devices_by_name = {device["name"]: device for device in sensor_devices}
+
+    # Add the leap data into the lip data
+    # so we know the type, model, and serial
+    for device in button_devices_by_id.values():
+        area = device.get("Area", {}).get("Name", "")
+        name = device["Name"]
+        leap_name = f"{area}_{name}"
+        device["leap_name"] = leap_name
+        leap_device_data = sensor_devices_by_name.get(leap_name)
+        if leap_device_data is None:
+            continue
+        for key in ("type", "model", "serial"):
+            val = leap_device_data.get(key)
+            if val is not None:
+                device[key] = val
+
+    _LOGGER.debug("Button Devices: %s", button_devices_by_id)
+    return button_devices_by_id
+
+
+async def _async_register_bridge_device(hass, config_entry_id, bridge_device):
+    """Register the bridge device in the device registry."""
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        name=bridge_device["name"],
+        manufacturer=MANUFACTURER,
+        config_entry_id=config_entry_id,
+        identifiers={(DOMAIN, bridge_device["serial"])},
+        model=f"{bridge_device['model']} ({bridge_device['type']})",
+    )
+
+
+async def _async_register_button_devices(
+    hass, config_entry_id, bridge_device, button_devices_by_id
+):
+    """Register button devices (Pico Remotes) in the device registry."""
+    device_registry = await dr.async_get_registry(hass)
+    button_devices_by_dr_id = {}
+
+    for device in button_devices_by_id.values():
+        if "serial" not in device:
+            continue
+
+        dr_device = device_registry.async_get_or_create(
+            name=device["leap_name"],
+            manufacturer=MANUFACTURER,
+            config_entry_id=config_entry_id,
+            identifiers={(DOMAIN, device["serial"])},
+            model=f"{device['model']} ({device['type']})",
+            via_device=(DOMAIN, bridge_device["serial"]),
+        )
+
+        button_devices_by_dr_id[dr_device.id] = device
+
+    return button_devices_by_dr_id
+
+
+@callback
+def _async_subscribe_pico_remote_events(hass, lip, button_devices_by_id):
```

Comment:
Just looking at this name led me to believe it was async (and should therefore be awaited).  Is it maybe better to remove `_async` from the name here to avoid confusion?

Reply:
Its decorated with `@callback` to indicate it is safe to run in async context (and if we ever pass it to hass.async_add_job it will run in async).